File: chart/src/main/java/se/tink/utils/CurrencyUtils.java
package se.tink.utils;

import android.util.SparseArray;
import com.google.common.base.Strings;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Currency;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class CurrencyUtils {

	private Locale locale;

	private static final int BILLION = 1000000000;
	private static final int MILLION = 1000000;
	private static final int THOUSAND = 1000;
	private static final int ZERO = 0;


	private static final Map<String, SparseArray<String>> UNITS = new HashMap<String, SparseArray<String>>();

	static {
		SparseArray<String> enUsMap = new SparseArray<String>();
		enUsMap.put(BILLION, "bn");
		enUsMap.put(MILLION, "mm");
		enUsMap.put(THOUSAND, "k");
		enUsMap.put(ZERO, "");

		SparseArray<String> svSeMap = new SparseArray<>();
		svSeMap.put(BILLION, "md");
		svSeMap.put(MILLION, "mn");
		svSeMap.put(THOUSAND, "t");
		svSeMap.put(ZERO, "");

		SparseArray<String> frFRMap = new SparseArray<>();
		frFRMap.put(BILLION, "md");
		frFRMap.put(MILLION, "mn");
		frFRMap.put(THOUSAND, "m");
		frFRMap.put(ZERO, "");

		SparseArray<String> defaultMap = new SparseArray<>();
		defaultMap.put(BILLION, "G");
		defaultMap.put(MILLION, "M");
		defaultMap.put(THOUSAND, "k");
		defaultMap.put(ZERO, "");

		SparseArray<String> nlNlMap = new SparseArray<>();
		nlNlMap.put(MILLION, "mln");
		nlNlMap.put(THOUSAND, "dzd");
		nlNlMap.put(ZERO, "");

		UNITS.put("en", enUsMap);
		UNITS.put("sv", svSeMap);
		UNITS.put("fr", frFRMap);
		UNITS.put("nl", nlNlMap);
		UNITS.put("", defaultMap);
	}

	private static CurrencyUtils instance;

	public static CurrencyUtils getInstance(Locale locale) {
		if (instance == null) {
			instance = new CurrencyUtils();
		}
		instance.setupLocale(locale);

		return instance;
	}

	private void setupLocale(Locale locale) {
		this.locale = locale;
	}

	public String formatCurrencyWithAmountSignAndWithoutSymbol(double amount, String currencyCode) {
		return formatCurrencyRound(amount, currencyCode, true, false);
	}

	public String formatCurrencyWithAmountSignAndWithSymbol(double amount, String currencyCode) {
		return formatCurrencyRound(amount, currencyCode, true, true);
	}

	public String formatCurrencyWithoutAmountSignAndWithoutSymbol(double amount,
		String currencyCode) {
		return formatCurrencyRound(amount, currencyCode, false, false);
	}

	public String formatCurrencyWithoutAmountSignAndWithSymbol(double amount, String currencyCode) {
		return formatCurrencyRound(amount, currencyCode, false, true);
	}

	private String formatCurrencyRound(double amount, String currencyCode, boolean withAmountSign,
		boolean useCurrencySymbol) {
		NumberFormat format = getDecimalFormat(currencyCode, 0);
		String formatted = format.format(Math.round(Math.abs(amount)));

		if (!useCurrencySymbol) {
			String symbol = ((DecimalFormat) format).getDecimalFormatSymbols().getCurrencySymbol();
			if (formatted.contains(symbol)) {
				formatted = formatted.replace(symbol, "");
			}
		}

		if (withAmountSign) {
			if (amount < 0) {
				formatted = "-" + formatted;
			}
		}

		return formatted;
	}

	public String formatShort(double value, String currencyCode) {
		Integer unit = null;
		String unitSign = "";
		String language = locale.getLanguage();

		SparseArray<String> localeUnits = UNITS.get(language);
		if (localeUnits == null) {
			localeUnits = UNITS.get("");
		}

		// Iterating in descending order since SparseArray does key ordering in ascending order
		for (int i = localeUnits.size() - 1; i >= 0; i--) {
			if (Double.compare(Math.abs(value), localeUnits.keyAt(i)) >= 0) {
				unit = localeUnits.keyAt(i);
				unitSign = localeUnits.get(unit);
				break;
			}
		}

		Currency currency = Currency.getInstance(currencyCode);

		String symbol = currency.getSymbol();
		if (symbol.equals("SEK")) {
			symbol = "kr";
		}

		if (Strings.isNullOrEmpty(unitSign)) {
			return String.format(locale, "%d %s", (int) value, symbol);
		}

		double valueInUnit = value / unit;

//		format.setMaximumIntegerDigits(3);

		boolean currencyLeading =
			currency.getSymbol().equals("$") || currency.getSymbol().equals("€") || currency
				.getSymbol().equals("£");

		int decimals = 0;
		// Check if value in unit is an integer.
		if (valueInUnit % 1 != 0) {
			decimals = 1;
		}

		if (currencyLeading) {
			return String.format(locale, "%s%s %s", symbol,
				formatAmountNoCurrency(valueInUnit, 0), unitSign);
		} else {

			return String
				.format(locale, "%s %s%s", formatAmountNoCurrency(valueInUnit, 0), unitSign,
					symbol);
		}
	}

	private String formatAmountNoCurrency(double amount, int decimals) {
		return getDecimalFormatNoCurrency(decimals).format(amount);
	}

	private NumberFormat getDecimalFormatNoCurrency(int decimals) {
		NumberFormat format = DecimalFormat.getInstance(locale);
		format.setMinimumFractionDigits(decimals);

		return format;
	}


	private String formatAmount(double amount, int decimals, String currencyCode) {
		return getDecimalFormat(currencyCode, decimals).format(amount);
	}

	private NumberFormat getDecimalFormat(String currencyCode, int decimals) {
		NumberFormat format;// = DecimalFormat.getCurrencyInstance(locale);
		if (currencyCode.equals("SEK")) {
			format = DecimalFormat.getCurrencyInstance(Locale.forLanguageTag("sv"));
		} else {
			format = DecimalFormat.getCurrencyInstance();
		}
		format.setCurrency(Currency.getInstance(currencyCode));
		format.setMinimumFractionDigits(decimals);

		return format;
	}


}

File: moneymanager-ui/src/main/java/com/tink/moneymanagerui/util/CurrencyUtils.kt
package com.tink.moneymanagerui.util

import com.tink.model.misc.Amount
import com.tink.model.misc.ExactNumber
import com.tink.moneymanagerui.collections.Currencies
import se.tink.commons.extensions.absValue
import se.tink.commons.extensions.doubleValue
import se.tink.commons.extensions.isBiggerThan
import se.tink.commons.extensions.isInteger
import se.tink.commons.extensions.isSmallerThan
import se.tink.commons.extensions.round
import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.Currency
import java.util.HashMap
import java.util.Locale
import kotlin.collections.set
import kotlin.math.roundToLong

object CurrencyUtils {
    // When using dynamic currency format:
    // if amount >= DYNAMIC_ROUNDING_THRESHOLD then rounded format is applied
    // else exact format is applied.
    private val DYNAMIC_ROUNDING_THRESHOLDS = DynamicRoundingThresholds
    private val ZERO = ExactNumber(0, 0)
    private val THOUSAND = ExactNumber(1000, 0)
    private val MILLION =
        ExactNumber(1000000, 0)
    private val BILLION =
        ExactNumber(1000000000, 0)
    private val UNITS: MutableMap<String, List<StringExactNumberPair>> =
        HashMap()

    private const val EXACT_DIGITS = 2

    private const val DEFAULT_CURRENCY_CODE = "EUR"

    fun formatCurrencyRound(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.ROUND or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN
        )
    }

    fun formatCurrencyRoundWithoutSign(amount: Amount): String {
        return formatCurrency(amount, CurrencyFormat.ROUND or CurrencyFormat.SYMBOL)
    }

    fun formatCurrencyExactWithoutSign(amount: Amount): String {
        return formatCurrency(amount, CurrencyFormat.EXACT or CurrencyFormat.SYMBOL)
    }

    fun formatCurrencyExact(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.EXACT or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN
        )
    }

    fun formatCurrencyExactIfNotInteger(amount: Amount, additionalFlags: Int = 0): String {
        return formatCurrency(
            amount,
            CurrencyFormat.EXACT_OR_INTEGER or CurrencyFormat.SYMBOL or additionalFlags
        )
    }

    fun formatCurrencyRoundWithoutAmountSign(amount: Amount): String {
        return formatCurrency(amount, CurrencyFormat.ROUND or CurrencyFormat.SYMBOL)
    }

    fun formatCurrencyWithAmountSign(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.DYNAMIC or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN
        )
    }

    fun formatCurrencyWithAmountSignExact(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.DYNAMIC or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN
                    or CurrencyFormat.EXACT
        )
    }

    fun formatCurrencyWithoutAmountSign(amount: Amount): String {
        return formatCurrency(amount, CurrencyFormat.DYNAMIC or CurrencyFormat.SYMBOL)
    }

    fun formatCurrencyWithExplicitPositive(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.DYNAMIC or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN,
            true
        )
    }

    fun formatCurrency(
        amount: Amount,
        currencyFormat: Int,
        explicitPositive: Boolean
    ): String {
        val absValue: ExactNumber = amount.value.absValue()
        var currencyCode = amount.currencyCode
        if (currencyCode.isBlank()) {
            currencyCode = DEFAULT_CURRENCY_CODE
        }
        var formatted: String
        formatted = if (currencyFormat and CurrencyFormat.ROUND == CurrencyFormat.ROUND) {
            formatAmountRound(absValue, currencyCode)
        } else if (currencyFormat and CurrencyFormat.EXACT == CurrencyFormat.EXACT) {
            formatAmountExact(absValue, currencyCode)
//        } else if (currencyFormat and CurrencyFormat.SHORT == CurrencyFormat.SHORT) {
//            formatShort(absValue, currencyCode)
        } else if(currencyFormat and CurrencyFormat.EXACT_OR_INTEGER == CurrencyFormat.EXACT_OR_INTEGER) {
            formatAmountExactIfNotInteger(absValue, currencyCode)
        } else { // CurrencyFormat.DYNAMIC
            if (absValue.toBigDecimal().toInt() < DYNAMIC_ROUNDING_THRESHOLDS.get(currencyCode) &&
                absValue.isBiggerThan(ZERO)
            ) {
                formatAmount(absValue, 2, currencyCode)
            } else {
                formatAmountRound(absValue, currencyCode)
            }
        }
        if (currencyFormat and CurrencyFormat.AMOUNT_SIGN == CurrencyFormat.AMOUNT_SIGN) {
            if (amount.value.isSmallerThan(ZERO)) {
                formatted = "-$formatted"
            } else if (explicitPositive) {
                formatted = "+$formatted"
            }
        }
        return formatted
    }

    fun formatCurrency(amount: Amount, currencyFormat: Int): String {
        return formatCurrency(amount, currencyFormat, false)
    }

    fun formatAmountRound(
        amount: ExactNumber,
        currencyCode: String?
    ): String {
        return formatAmount(amount, 0, currencyCode)
    }

    fun formatAmountRoundWithoutCurrencySymbol(amount: Double, currency: String?): String {
        return formatAmountRound(amount, currency, false)
    }

    @JvmStatic
    fun formatAmountRoundWithCurrencySymbol(amount: Double, currency: String): String {
        return formatAmountRound(amount, currency, true)
    }

    private fun formatAmountRound(
        amount: Double,
        currency: String?,
        useCurrencySymbol: Boolean
    ): String {
        val format = getDecimalFormat(currency, 0)
        var formatted = format.format(amount.roundToLong())
        if (!useCurrencySymbol) {
            val symbol =
                (format as DecimalFormat).decimalFormatSymbols.currencySymbol
            if (formatted.contains(symbol)) {
                //Replace the currency symbol and the surrounding space, if it's before or after the symbol, or no space.
                formatted = formatted.replace(symbol + "\\s".toRegex(), "")
                formatted = formatted.replace("\\s" + symbol.toRegex(), "")
                formatted = formatted.replace(symbol.toRegex(), "")
            }
        }
        return formatted
    }

    private fun formatAmount(
        amount: ExactNumber,
        decimals: Int,
        currencyCode: String?
    ): String {
        return getDecimalFormat(currencyCode, decimals)
            .format(amount.round(decimals).toBigDecimal())
    }

    private fun formatIntegerAmount(
        amount: ExactNumber,
        currencyCode: String
    ): String {
        val currencySymbol = try {
            Currency.getInstance(currencyCode).symbol
        } catch (exception: Exception) {
            Currency.getInstance(Currencies.getSharedInstance().defaultCurrencyCode).symbol
        }
        return "$currencySymbol${amount.toBigDecimal().toLong()}"
    }

    // TODO: Remove deprecated Currencies usage
    private fun getDecimalFormat(
        currencyCode: String?,
        decimals: Int
    ): NumberFormat {
        var currencyCode = currencyCode
        var locale = Locale.getDefault()
        val userProfile =
            Currencies.getSharedInstance().userProfile
        if (userProfile != null) {
            val localeCode = userProfile.locale
            val splits = localeCode.split("_").toTypedArray()
            locale = Locale(splits[0], userProfile.market)
        }
        if (currencyCode == null) {
            currencyCode = Currencies.getSharedInstance().defaultCurrencyCode
        }
        val format = DecimalFormat.getCurrencyInstance(locale)
        if (!currencyCode.isEmpty()) {
            format.currency = Currency.getInstance(currencyCode)
        }
        format.minimumFractionDigits = decimals
        format.maximumFractionDigits = decimals
        return format
    }

    fun formatCurrencyExactWithExplicitPositive(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.EXACT or CurrencyFormat.SYMBOL or CurrencyFormat.AMOUNT_SIGN,
            true
        )
    }

    fun formatCurrencyExactWithoutSymbol(amount: Amount): String {
        return formatCurrency(
            amount,
            CurrencyFormat.EXACT or CurrencyFormat.AMOUNT_SIGN
        )
    }

    fun formatAmountExactWithoutCurrencySymbol(amount: Double, currency: String?): String {
        return formatAmount(amount, 2, currency, false);
    }

    fun formatCurrencyExactWithoutSignAndSymbol(amount: Amount, currency: String?): String {
        return formatAmount(amount.value.absValue().doubleValue(), 2, currency, false)
    }

    fun formatAmountExactWithCurrencySymbol(amount: Double, currency: String?): String {
        return formatAmount(amount, 2, currency, true);
    }

    private fun formatAmountExact(
        amount: ExactNumber,
        currencyCode: String
    ): String {
        return formatAmount(amount, EXACT_DIGITS, currencyCode)
    }

    private fun formatAmountExactIfNotInteger(
        amount: ExactNumber,
        currencyCode: String
    ): String {
        return if (amount.isInteger()) {
            formatIntegerAmount(amount, currencyCode)
        } else {
            formatAmountExact(amount, currencyCode)
        }
    }

    private fun formatAmount(amount: Double, decimals: Int, currency: String?, useCurrencySymbol: Boolean): String {
        val format = getDecimalFormat(currency, decimals)
        var formatted = format.format(amount)
        if (!useCurrencySymbol) {
            val symbol = (format as DecimalFormat).decimalFormatSymbols.currencySymbol
            if (formatted.contains(symbol)) {
                //Replace the currency symbol and the surrounding space, if it's before or after the symbol, or no space.
                formatted = formatted.replace(symbol + "\\s".toRegex(), "")
                formatted = formatted.replace("\\s" + symbol.toRegex(), "")
                formatted = formatted.replace(symbol.toRegex(), "")
            }
        }
        return formatted
    }

    private class StringExactNumberPair(
        val s: String,
        val e: ExactNumber
    )

    object CurrencyFormat {
        // Value dependent format (exact if very small, rounded otherwise).
        const val NONE = 0x0
        const val SYMBOL = 0x1 // Include currency symbol.
        const val SHORT = 0x2
        const val EXACT = 0x4
        const val ROUND = 0x8
        const val DYNAMIC = 0x10
        const val VERY_SHORT = 0x16
        const val AMOUNT_SIGN = 0x20 // Include amount sign
        const val EXACT_OR_INTEGER = 0x40 // Keep as integer if no decimals, otherwise exact
    }

    init {
        val enUsMap =
            listOf(
                    StringExactNumberPair("", ZERO),
                    StringExactNumberPair("k", THOUSAND),
                    StringExactNumberPair("mm", MILLION),
                    StringExactNumberPair("bn", BILLION)
                )
        val svSeMap =
            listOf(
                    StringExactNumberPair("", ZERO),
                    StringExactNumberPair("t", THOUSAND),
                    StringExactNumberPair("mn", MILLION),
                    StringExactNumberPair("md", BILLION)
                )
        val frFRMap =
            listOf(
                    StringExactNumberPair("", ZERO),
                    StringExactNumberPair("m", THOUSAND),
                    StringExactNumberPair("mn", MILLION),
                    StringExactNumberPair("md", BILLION)
                )
        val defaultMap =
            listOf(
                    StringExactNumberPair("", ZERO),
                    StringExactNumberPair("k", THOUSAND),
                    StringExactNumberPair("M", MILLION),
                    StringExactNumberPair("G", BILLION)
                )
        val nlNlMap =
            listOf(
                    StringExactNumberPair("", ZERO),
                    StringExactNumberPair("dzd", THOUSAND),
                    StringExactNumberPair("mln", MILLION),
                    StringExactNumberPair("mjd", BILLION)
                )
        UNITS["en"] = enUsMap
        UNITS["sv"] = svSeMap
        UNITS["fr"] = frFRMap
        UNITS[""] = defaultMap
        UNITS["nl"] = nlNlMap
    }
}
